src/corecel/data/LdgIterator.hh

```diff
@@ -0,0 +1,253 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/data/LdgIterator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <type_traits>
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/math/Algorithms.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+
+/*!
+ * Reads a value T using __ldg builtin and return a copy of it
+ */
+template<class T, typename = void>
+struct LDGLoad
+{
+    using value_type = T;
+    using pointer = std::add_pointer_t<value_type const>;
+    using reference = value_type;
+
+    CELER_CONSTEXPR_FUNCTION static reference read(pointer p)
```

Comment:
Since this isn't trying to conform to any existing interface, perhaps `LdgLoader` with `operator()`? Accessing with an instance could make sense if we extend LDG loading to "read member data from a struct" (e.g., for struct `T`, the loader could hold `U T::*` as class data and read member data from it using ldg?) I think this and the traits class should be in `data/detail/LdgIteratorImpl.hh` since they're more than a couple of lines.

Reply:
I think it's easier to make this class stateless otherwise the LdgIterator has to make sure they both stay consistent.